polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala

```diff
@@ -137,87 +82,121 @@ class PythonInterpreter(val symbolTable: RuntimeSymbolTable) extends LanguageKer
-      jep.eval("globals().update(__polynote_locals__)")
-      val newDecls = jep.getValue("list(__polynote_locals__.keys())", classOf[java.util.List[String]]).asScala.toList
+          jep.set("__polynote_code__", maybeModifiedCode)
+          jep.eval("exec(__polynote_code__, __polynote_globals__, __polynote_locals__)\n")
+          jep.eval("globals().update(__polynote_locals__)") // still need to update globals so we can eval stuff
+          val newDecls = jep.getValue("list(__polynote_locals__.keys())", classOf[java.util.List[String]]).asScala.toList
 
       // TODO: We talked about potentially creating an `Out` map, inspired by ipython, instead of these resCell# vars
-      val maybeOutput = if (lastIsExpr) {
+      val (maybeOutput, maybeReturns) = if (lastIsExpr) {
         val hasLastValue = jep.getValue(s"$resultName != None", classOf[java.lang.Boolean])
         if (hasLastValue) {
-          Option(Output("text/plain; rel=decl; lang=python", s"$resultName = ${jep.getValue(s"repr($resultName)", classOf[String])}"))
-        } else None
-      } else None
+          (
+            Option(Output("text/plain; rel=decl; lang=python", s"$resultName = ${jep.getValue(s"repr(${newDecls.last})", classOf[String])}")),
+            Option(globalInfo.RuntimeValue(resultName, jep.getValue(resultName), Some(this), cell))
+          )
+        } else (None, None)
+      } else (None, None)
 
-      (getPyResults(newDecls, cell), maybeOutput)
+          val symbolDict = getPyResults(newDecls, cell).filterNot { case (k, v) => k == resultName || v == null }
 
-    }.flatMap {
-      case (resultDict, maybeOutput) =>
-        val symbolPub = symbolTable.publishAll(resultDict.filterNot(_._2.value == null).values.toList)
+          val symbols = symbolDict.map {
+            case (name, value) => name -> globalInfo.RuntimeValue(name, value, Some(this), cell)
+          }
 
-        maybeOutput
-          .map(o => out.enqueue1(o))
-          .getOrElse(IO.unit)
-          .flatMap( _ => symbolPub)
+          (maybeOutput, symbols, maybeReturns)
+        }
+      }
 
-    }.flatMap {
-      _ => IO(stdout.flush())
+      Queue.unbounded[IO, Option[Result]].flatMap { maybeResultQ =>
+        val eval = for {
+          globalDict <- symbolsToPythonDict(runtimeContext.visibleSymbols)
+          resultQ = new EnqueueSome(maybeResultQ)
+          stdout = new PythonDisplayHook(resultQ)
+          execResult <- exec(stdout, globalDict)
+          (maybeOutput, symbols, maybeReturns) = execResult
+          _ <- IO(stdout.flush())
+          _ <- Stream.emits(maybeOutput.toSeq).to(resultQ.enqueue).compile.toVector
+        } yield {
+
+
+          // TODO: this IO isn't really useful. How can we return the Stream and IO before we are ready?
+          val newCtx = IO(RuntimeContext(cell, globalInfo, Option(runtimeContext), symbols, None, maybeReturns))
+          (maybeResultQ.dequeue.unNoneTerminate, newCtx)
```

Comment:
So I think that just wrapping an `IO` around this doesn't actually achieve what we want - all the processing will be fully complete before this is returned, right? At this point I think I might need some `IO`/stream pointers...